DEVELOPMENT.md

```diff
@@ -0,0 +1,473 @@
+Environment-specific feature flags are supported for both the api and frontend. Flagged features are assumed to be "off" unless the flag exists (and the value is truthy), thus feature flag conditions should always check for the presence or truthiness of the flag, not for it's absence. Environment variables for feature flags *MUST* be prefixed by `FEATURE_`, ex. `FEATURE_BRING_THE_AWESOME` or `FEATURE_BRING_DA_RUCKUS`.
+
+### Current available features
+
+`FEATURE_AUTH_GITHUB`: Used in [`api/controllers/main.js`](./api/controllers/main.js) to enable authentication with Github oauth.
+`FEATURE_AUTH_UAA`: Used in [`api/controllers/main.js`](./api/controllers/main.js) to enable authentication with the cloud.gov UAA.
+
+### Api feature flags
+Api feature flags are evaluated at *runtime* and should be created explicitly in the code before the corresponding environment variable can be used. Example:
+
+Given environment variable `FEATURE_AWESOME_SAUCE='true'`
+
+1. Add the flag to the known flags:
+```js
+// api/features.js
+const Flags = {
+  //...
+  FEATURE_AWESOME_SAUCE: 'FEATURE_AWESOME_SAUCE',
+}
+```
+2. Check if the feature is enabled:
+```js
+// some code in a child folder of /api
+const Features = require('../features');
+
+if (Features.enabled(Features.Flags.FEATURE_AWESOME_SAUCE)) {
+  doSomething();
+}
+```
+
+### Frontend feature flags
+Frontend feature flags are evaluated at *compile* time NOT at runtime, resulting in unused codepaths being exluded from the built code. Environment variables with the `FEATURE_` prefix are available globally within the codebase at compile time. This magic is performed by `webpack.DefinePlugin` and minification.
+
+
+Example:
+
+Given environment variable `FEATURE_AWESOME_SAUCE='true'`
+
+```js
+if (FEATURE_AWESOME_SAUCE === 'true') {
+  doSomething();
+}
+```
+
+results in
+
+```js
+doSomething();
+```
+
+.
+
+## Build Logs
+
+Build logs are streamed directly to the database in realtime so they are immediately viewable by customers in the UI. Every night, build logs for completed builds for the day are archived in an S3 bucket and deleted from the database. This prevents the database from growing unreasonably. The S3 bucket (with assistance from cloud.gov) is configured with a lifecycle policy that expires the logs after 180 days in accordance with [cloud.gov's policies on log retention](https://cloud.gov/docs/deployment/logs/#web-based-logs-with-historic-log-data).
+
+Lifecycle policy:
+```
+{
+  Bucket: <bucket>,
+  LifecycleConfiguration: {
+    Rules: [
+      {
+        Expiration: {
+          Days: 180
+        },
+        ID: "Archive all objects 180 days after creation"
+        Status: "Enabled",
+      }
+    ]
+  }
+}
+```
+
+## Commits, Releases, and Deployment
+
+All commit messages on the default branch should follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) specification. In addition, an associated ticket number should be added where relevant. Examples:
```

Comment:
It's not that unusual for one of our commits to need to reference a ticket on one of our other GitHub repos. Should this perhaps clarify what those should look like?

Reply:
For example, the runbook I just did was a ticket on this repo, but the implementation was a commit to a different repo.